File: easy/111. Minimum Depth of Binary Tree/111. Minimum Depth of Binary Tree.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

function minDepth(root: TreeNode | null): number {
    if (!root) return 0

    const queue = [root]

    let depth = 0

    while (queue.length > 0) {
        const levelSize = queue.length
        depth += 1

        for (let i = 0; i < levelSize; i++) {
            const currentNode = queue.shift()!

            if (!currentNode.left && !currentNode.right) {
                return depth
            } else {
                if (currentNode.left) queue.push(currentNode.left)
                if (currentNode.right) queue.push(currentNode.right)
            }
        }
    }

    return queue.length
};

export { }

File: easy/111. Minimum Depth of Binary Tree/111. Minimum Depth of Binary Tree.py
from collections import deque
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        depth: int = 0
        queue: deque[TreeNode] = deque([root])

        while len(queue) > 0:
            level_size = len(queue)
            depth += 1

            for _ in range(level_size):
                curr_node = queue.popleft()

                if not curr_node.right and not curr_node.left:
                    return depth

                if curr_node.left:
                    queue.append(curr_node.left)
                if curr_node.right:
                    queue.append(curr_node.right)

        return len(queue)
